File: 0104_Maximum_Depth_of_Binary_Tre/0104_myans.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traceNode(self, parent: Optional[TreeNode], level: int) -> int:
        if parent == None:
            return level
        
        leftlevel = self.traceNode(parent.left, level+1)
        rightlevel = self.traceNode(parent.right, level+1)

        return max(leftlevel, rightlevel)

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        level = 0
        return self.traceNode(root, level)
        

File: 0104_Maximum_Depth_of_Binary_Tre/0104_myans.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
    int traceNode(TreeNode *parent, int level)
    {
        int leftDepth = level;
        int rightDepth = level;

        if (parent->left != nullptr)
            leftDepth = traceNode(parent->left, level + 1);
        if (parent->right != nullptr)
            rightDepth = traceNode(parent->right, level + 1);

        return max(leftDepth, rightDepth);
    }

public:
    int maxDepth(TreeNode *root)
    {
        if (root == nullptr)
            return 0;
        int level = 1;
        return traceNode(root, level);
    }
};